drivers: wifi: siwx91x: Add power-save support

ArunmaniAlagarsamy2710 opened the pull request:
Added power-save mode support to optimize station wake-up timing
and reduce power consumption. The device currently supports
only legacy power-save mode.

Implemented power-save features:
- ` ps ON/OFF`	     : By default, the device operates in
		                fast PSP mode.
- `ps_listen_interval `: Moves to DTIM wakeup if set to zero.
- `ps_exit_strategy   `: Updates mode if enabled; otherwise, follows
		                   the configured exit strategy when power-save
		                   is enabled.
- `ps_wakeup_mode     `: Configures the wake-up behavior.
- `ps_timeout         `: Defines the timeout duration for power-save mode.

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +750,179 @@ static int siwx91x_dev_init(const struct device *dev)
+		LOG_ERR("Wi-Fi not in station mode");
+		return -EIO;
+	}
+
```
check for interface is disabled

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +750,179 @@ static int siwx91x_dev_init(const struct device *dev)
 	return 0;
 }
 
+sl_si91x_performance_profile_t req_ps_mode;
```
comment for this variable usage

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +750,179 @@ static int siwx91x_dev_init(const struct device *dev)
+			return -EINVAL;
+		}
+
+		if (sl_ps_profile.profile != HIGH_PERFORMANCE) {
```
Better keep this way

if (sl_ps_profile.profile == HIGH_PERFORMANCE) {
    return 0;
}

sl_ps_profile.profile = req_ps_mode;

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +750,179 @@ static int siwx91x_dev_init(const struct device *dev)
+		break;
+	case WIFI_PS_PARAM_LISTEN_INTERVAL:
+		/* why dont use the local macro same as MAX for MIN to maintain consistency? */
+		if (params->listen_interval < WIFI_LISTEN_INTERVAL_MIN ||
```
can skip this check WIFI_LISTEN_INTERVAL_MIN, as it is 0

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +750,179 @@ static int siwx91x_dev_init(const struct device *dev)
+
+	switch (params->type) {
+	case WIFI_PS_PARAM_STATE:
+		if (params->enabled) {
```
(params->enabled == WIFI_PS_ENABLED)

NiteshKShah commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -49,4 +49,12 @@ config NET_MGMT_EVENT_QUEUE_SIZE
 # This device supports filtering scan results for only one SSID.
 config WIFI_MGMT_SCAN_SSID_FILT_MAX
 	default 1
+
```
remove this for now.

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +750,179 @@ static int siwx91x_dev_init(const struct device *dev)
+			} else {
+				return -EAGAIN;
+			}
+		} else {
```
(params->enabled == WIFI_PS_DISABLED)

krish2718 commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -709,6 +713,193 @@ static int siwx91x_stats(const struct device *dev, struct net_stats_wifi *stats)
+					sl_ps_profile.profile = req_ps_mode;
+				}
+			} else {
+				return -EAGAIN;
```
did you mean `-EALREADY`?

jhedberg commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -709,6 +712,237 @@ static int siwx91x_stats(const struct device *dev, struct net_stats_wifi *stats)
+	get_wifi_current_performance_profile(&sl_ps_profile);
+
+	switch (params->type) {
+	case WIFI_PS_PARAM_STATE:
+		if (params->enabled == WIFI_PS_ENABLED) {
+			if (sl_ps_profile.profile == HIGH_PERFORMANCE) {
+				if (siwx91x_ps_cfg.req_ps_mode == HIGH_PERFORMANCE) {
+					sl_ps_profile.profile = ASSOCIATED_POWER_SAVE_LOW_LATENCY;
+				} else {
+					sl_ps_profile.profile = siwx91x_ps_cfg.req_ps_mode;
+				}
+			} else {
+				return -EALREADY;
+			}
+
+			if (sl_ps_profile.dtim_aligned_type == 0) {
+				beacon_interval = siwx91x_get_beacon_interval(dev);
+				if (beacon_interval < 0) {
+					params->fail_reason = WIFI_PS_PARAM_FAIL_DEVICE_NOT_CONNECTED;
```
The compliance check is failing because of the long (> 100) lines. To me it looks like this function could use some refactoring into smaller helper functions, so that you don't need to do so deep indentations.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -709,6 +712,270 @@ static int siwx91x_stats(const struct device *dev, struct net_stats_wifi *stats)
+		if (beacon_interval > 0) {
+			/* converts MSEC to beacon unit */
+			config->ps_params.listen_interval =
+				(sl_ps_profile.listen_interval / beacon_interval);
```
Parentheses are useless.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -709,6 +712,270 @@ static int siwx91x_stats(const struct device *dev, struct net_stats_wifi *stats)
+
+	if (sl_ps_profile.dtim_aligned_type == 1) {
+		config->ps_params.wakeup_mode = WIFI_PS_WAKEUP_MODE_DTIM;
+	} else if (sl_ps_profile.dtim_aligned_type == 0) {
```
If you consider `dtim_aligned_type` is a boolean, test it as a boolean:
```
	if (sl_ps_profile.dtim_aligned_type) {
            ...
	} else {
            ...
        }
```

jerome-pouiller commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -49,4 +49,11 @@ config NET_MGMT_EVENT_QUEUE_SIZE
+	help
+	  Enable this option to allow the device to configure
+	  Enhanced Maximum Power Save (PSP) support during
+	  device initialization.
```
This option only makes sense if `WIFI_SILABS_SIWX91X` is selected.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -709,6 +712,270 @@ static int siwx91x_stats(const struct device *dev, struct net_stats_wifi *stats)
+		return -EINVAL;
+	}
+
+	return sys_get_le16(sl_stat.beacon_interval);
```
Until now, we didn't considered the host CPU could be big endian. I believe it does not make sense to do it for `sl_stat.beacon_interval`.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -709,6 +712,270 @@ static int siwx91x_stats(const struct device *dev, struct net_stats_wifi *stats)
+		*listen_interval = SIWX91X_PS_MAX_LISTEN_INTERVAL_MS;
+	}
+
+	return 0;
```
I suggest to return the result. Hence, you will be consistent with `siwx91x_get_beacon_interval()`.

ArunmaniAlagarsamy2710 replied:
I agree with the point on consistency with siwx91x_get_beacon_interval(). However, since the function name is now changed to siwx91x_set_listen_interval(), I believe it makes sense to explicitly set the listen interval value within the function. Would this be a reasonable approach in this case?

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -709,6 +712,270 @@ static int siwx91x_stats(const struct device *dev, struct net_stats_wifi *stats)
+
+	__ASSERT(sl_ps_profile, "sl_ps_profile cannot be NULL");
+
+	switch(state) {
```
Missing space after `switch`.

ArunmaniAlagarsamy2710 commented:
Rebased to reduce the CI failures

ArunmaniAlagarsamy2710 committed:
drivers: wifi: siwx91x: Add power-save support

Added power-save mode support to optimize station wake-up timing
and reduce power consumption. The device currently supports
only legacy power-save mode.

Implemented power-save features:
- ps ON/OFF	     : By default, the device operates in
		       fast PSP mode.
- ps_exit_strategy   : Updates mode if enabled; otherwise, follows
		       the configured exit strategy when power-save
		       is enabled.
- ps_wakeup_mode     : Configures the wake-up behavior.
- ps_timeout         : Defines the timeout duration for power-save mode.

Signed-off-by: Arunmani Alagarsamy <arunmani.a@silabs.com>

ArunmaniAlagarsamy2710 committed:
drivers: wifi: siwx91x: Add listen interval config

Adds support to configure listen interval in power save mode.
Zephyr sets listen interval based on beacon units,
while SDK expects it in ms. The setting takes effect after
connection. If set to 0, device switches to DTIM wakeup mode.

Signed-off-by: Arunmani Alagarsamy <arunmani.a@silabs.com>

ArunmaniAlagarsamy2710 commented:
Rebased to resolve conflicts

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -755,11 +755,72 @@ static int siwx91x_get_version(const struct device *dev, struct wifi_version *pa
+static int siwx91x_convert_li_bcn_unit_to_ms(uint16_t bcn_li)
+{
+	uint16_t beacon_interval;
+	uint16_t ms_li;
```
Why `uint16_t`? Unless you have good reason use `int`.

jerome-pouiller replied:
The variable names lack of consistency in the function. I suggest:
   - `listen_interval`
   - `beacon_interval`
   - `listen_interval_ms`

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -755,11 +755,72 @@ static int siwx91x_get_version(const struct device *dev, struct wifi_version *pa
+		return -EAGAIN;
+	}
+
+	ms_li = bcn_li * beacon_interval;
```
If you want to use `uint16_t`, you have to add a comment explaining why you guarantee there is no overflow here.

jerome-pouiller replied:
It seems the units of beacon_interval are TUs. Are you sure you are calculating milliseconds here?

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -755,11 +755,72 @@ static int siwx91x_get_version(const struct device *dev, struct wifi_version *pa
+		return ms_li;
+	}
+
+	sl_ps_profile->listen_interval = ms_li;
```
Wiseconnect documentation says:
```
  uint16_t listen_interval;  ///< Listen interval in beacon intervals.
```

ArunmaniAlagarsamy2710 replied:
This is wrong documentation, Firmware is dealing the listen interval in ms

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -755,11 +755,72 @@ static int siwx91x_get_version(const struct device *dev, struct wifi_version *pa
+	if (ms_li > SIWX91X_PS_MAX_LISTEN_INTERVAL_MS) {
+		LOG_WRN("Listen interval exceeded, capping to %d",
+			SIWX91X_PS_MAX_LISTEN_INTERVAL_MS);
+		ms_li = SIWX91X_PS_MAX_LISTEN_INTERVAL_MS;
```
Where this constraint come from?

ArunmaniAlagarsamy2710 replied:
This is 917 hardware limitation

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -755,11 +755,72 @@ static int siwx91x_get_version(const struct device *dev, struct wifi_version *pa
+	if (sidev->state != WIFI_STATE_COMPLETED) {
+		LOG_WRN("Device not connected, listen interval reflect after connection");
+		return -EINVAL;
+	}
```
Take care of your design. I assume the job of this function is to fill `sl_ps_profile`. The state of the device is checked by the callers.

In addition, this function does not not "schedule" anything. It just fills `sl_ps_profile` with proper values.

Probably you want:
```
int siwx91x_set_listen_interval(sl_wifi_performance_profile_t *sl_ps_profile, uint16_t bcn_li);
```

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -755,11 +755,72 @@ static int siwx91x_get_version(const struct device *dev, struct wifi_version *pa
-				     sl_si91x_performance_profile_t exit_strategy_mode)
+				     sl_si91x_performance_profile_t exit_strategy_mode,
+				     uint16_t bcn_li)
 {
```
Same comment than for `siwx91x_schedule_listen_interval_wakeup()`.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -795,6 +864,7 @@ static int siwx91x_set_power_save(const struct device *dev, struct wifi_ps_param
 	struct siwx91x_dev *sidev = dev->data;
 	sl_wifi_performance_profile_t *sl_ps_profile = &sidev->sl_ps_profile;
 	sl_wifi_interface_t interface;
+	static uint16_t bcn_li;
```
Please (really) avoid global variables.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -816,7 +886,8 @@ static int siwx91x_set_power_save(const struct device *dev, struct wifi_ps_param
 	case WIFI_PS_PARAM_STATE:
-		status = siwx91x_configure_ps_mode(dev, params->enabled, exit_strategy_mode);
+		status = siwx91x_configure_ps_mode(dev, params->enabled, exit_strategy_mode,
+						   bcn_li);
```
Value of may be `bcn_li` is undefined.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -73,31 +73,6 @@ static int siwx91x_map_ap_security(enum wifi_security_type security)
-	siwx91x_on_join_ipv6(sidev);
-
-	return 0;
-}
```
Could you relocate this function in a separate commit so the review will be easier?

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -914,13 +892,14 @@ static int siwx91x_set_power_save(const struct device *dev, struct wifi_ps_param
-				if (status < 0) {
-					return status;
+				if (status == -EINVAL) {
+					sl_ps_profile->profile = HIGH_PERFORMANCE;
```
1. Is it related to this commit?
2. `EINVAL` means "Invalid argument". This is a bit weird to fix this error by changing the profile.
3. What happens now if `status < 0 &&  status != -EINVAL` ?

jerome-pouiller reviewed (changes requested):
This code is still super difficult to understand.

ArunmaniAlagarsamy2710 committed:
drivers: wifi: siwx91x: Relocate siwx91x_on_join

Move the siwx91x_on_join() function to make it more accessible
for power save feature integration. No functional changes.

Signed-off-by: Arunmani Alagarsamy <arunmani.a@silabs.com>

ArunmaniAlagarsamy2710 committed:
drivers: wifi: siwx91x: Support pre-connection listen interval wakeup

Allow users to enable listen interval wakeup mode before connection.
Power save is automatically activated after connection if configured
beforehand.

Signed-off-by: Arunmani Alagarsamy <arunmani.a@silabs.com>

ArunmaniAlagarsamy2710 commented:
Rebased to resolve conflicts

ArunmaniAlagarsamy2710 commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -753,6 +755,203 @@ static int siwx91x_get_version(const struct device *dev, struct wifi_version *pa
+
+static int siwx91x_set_power_save(const struct device *dev, struct wifi_ps_params *params)
+{
+	static sl_si91x_performance_profile_t exit_strategy_mode = { HIGH_PERFORMANCE };
```
Actually, we need the `exit_strategy_mode` variable to store the power save profile in case the user configures the exit strategy before the power save enabled. If we directly use `sl_ps_profile->profile`, power save would be enabled immediately even if the user hasn't explicitly enabled it.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -1039,6 +1042,224 @@ static int siwx91x_get_version(const struct device *dev, struct wifi_version *pa
+		if (sidev->exit_strategy_mode == HIGH_PERFORMANCE) {
+			sl_ps_profile->profile = ASSOCIATED_POWER_SAVE_LOW_LATENCY;
+		} else {
+			sl_ps_profile->profile = sidev->exit_strategy_mode;
```
If you change the default value of `exit_strategy_mode` into `ASSOCIATED_POWER_SAVE_LOW_LATENCY`, you can simplify the statement above.

Then `exit_strategy_mode` will only contains `ASSOCIATED_POWER_SAVE_LOW_LATENCY` or `ASSOCIATED_POWER_SAVE`, so it will be easier to understand.

ArunmaniAlagarsamy2710 committed:
drivers: wifi: siwx91x: Add helper to check device connection state

Introduced a dedicated function to check if the device is connected
to a Wi-Fi network. This helps simplify connection state handling
in the driver

Signed-off-by: Arunmani Alagarsamy <arunmani.a@silabs.com>

ArunmaniAlagarsamy2710 commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -910,12 +976,29 @@ static int siwx91x_set_power_save(const struct device *dev, struct wifi_ps_param
 		break;
+	case WIFI_PS_PARAM_LISTEN_INTERVAL:
+		sidev->bcn_listen_interval = params->listen_interval;
+		return 0;
```
Yes. For the listen interval, power save is enabled only after the device is connected. This value is stored in advance in case the user configures it before connection. Once the device is connected, the listen interval cannot be modified, so if power save is already active, the new value won't take effect.

jerome-pouiller replied:
Probably we want to catch the error?

ArunmaniAlagarsamy2710 commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -1039,6 +1042,224 @@ static int siwx91x_get_version(const struct device *dev, struct wifi_version *pa
+		break;
+	case WIFI_PS_PARAM_EXIT_STRATEGY:
+		status = siwx91x_set_ps_param_exit_strategy(sidev, params, sl_ps_profile);
+		if (status != SIWX91X_SET_PS) {
```
In this case, calling `sl_wifi_set_performance_profile()` unconditionally would not be ideal. It actually disables and re-enables power save each time, even if the profile hasn't changed.

ArunmaniAlagarsamy2710 commented:
Rebased to resolve conflicts

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -1326,12 +1326,22 @@ static void siwx91x_set_dtim_ps_profile(sl_wifi_performance_profile_t *sl_ps_pro
-	return siwx91x_try_set_listen_interval(sidev, sl_ps_profile, sidev->bcn_listen_interval);
+	status = siwx91x_try_set_listen_interval(sidev, sl_ps_profile, sidev->bcn_listen_interval);
+	if (status == -EACCES) {
+		sidev->listen_interval_wakeup = true;
```
Please do not introduce unexpected side effects.

ArunmaniAlagarsamy2710 commented:
superseded by https://github.com/zephyrproject-rtos/zephyr/pull/90284
